### PR TITLE
wang2018dynamic_1+zayne-sprague: Initial Draft

### PR DESCRIPTION
**Pull request by Zayne-sprague**
I need to add the formal definition of EdgeConv in Latex, but I also wonder if I should reduce the size of the images.  I took the via screen shots on my monitor because the ones online were too small, but now they are suuuuuper large (but maybe that's a good thing for presenting?)

### CODE REVIEW

**Comment by jaywhang** [comment]
> I need to add the formal definition of EdgeConv in Latex, but I also wonder if I should reduce the size of the images. I took the via screen shots on my monitor because the ones online were too small, but now they are suuuuuper large (but maybe that's a good thing for presenting?)

You can control the width of an image by manually inserting the image via html:
```
<p align="center">
  <img width="900" alt="Some alt text" src="path_to_image.png">
</p>
```

File: _summaries/wang2018dynamic_1.md [hunk 1]
```diff
@@ -1,19 +1,95 @@
...
+## The Core Idea of Dynamic Graph CNN
+
+- By utilizing edge features of nearby vertices connecting to another vertex, we can apply CNN operations on them computing local and global feature maps for the point cloud that is invariant to permutation, has local structure, and is easy to implement.
+
+## What are edges in a point cloud?
+
+- Graphs are defined by a local neighbourhood of vertices found by using KNN (which points are closest to the current vertex in question)
+
+- Edges are constructed by drawing directed edges to each vertex in the local graph, including edges pointing to the same vertex.
+    - This should sound familiar -- it's similar to self-attention from transformers
```
Comment (jaywhang) [inline _summaries/wang2018dynamic_1.md:20]: I'm having difficulty seeing the parallel to self-attention.  Specifically, in what way the input vectors (the `x_i`'s and `x_j`'s in this paper) are transformed into key/query vectors, and how do they "interact"?  For example, eq. 8 (and the equations before it that describe previous methods) shows that there is no direct dot product between the feature vectors.  So it feels more analogous to deformable convolution rather than self-attention.

File: _summaries/wang2018dynamic_1.md [hunk 2]
```diff
@@ -1,19 +1,95 @@
...
+## How do we calculate the values of these edges (how is it realized technically)
+
+- These edges are used as features for the DGCNN by running the vertices through a fully connected layer (or layers) multiple times
+
+- The feature for a vertex is an aggregation function applied to all the edge features (the convolution operation)
+
+- You can change the non-linearity and the aggregation function in different ways to achieve various effects.
+
+![Edge Conv](wang2018dynamic_1c.png)
+  - Look at how $$e_{ij}$$ is computed, you take both vertices and feed their features (in this case X,Y,Z coordinates) into linear layers that will be convoled over later
```
Comment (jaywhang) [inline _summaries/wang2018dynamic_1.md:33]: I think this diagram either requires more explanation, or could be replaced with something else.  I'm thinking maybe Table 1 from the paper.  Once you add the formal definition of EdgeConv, I think Table 1 is pretty easy to understand, and it really clearly shows in way sense EdgeConv is "both global and local".  Another added benefit is that the edge function's dependence only on `x_i` or `x_j`  for PointNet variants really shows how localized PointNet operations are.

File: _summaries/wang2018dynamic_1.md [hunk 3]
```diff
@@ -1,19 +1,95 @@
...
+- Once EdgeConv has ran on the point cloud, you now have a bunch of new values for the point cloud that have been mutated.
+
+- Empirically, the authors found that it is beneficial to recompute the K nearest neighbours after each edge conv.
+
+- This leads to points in the cloud becoming closer due to their feature space similarity not despite their input proximity!
+
+- This recomputation of the K nearest neighbour is a reconstruction of the graph and thus influenced the authors to name their method "Dynamic Graph CNN"
+
+
+## How is this different from PointNet & PointNet++
```
Comment (jaywhang) [inline _summaries/wang2018dynamic_1.md:50]: Very helpful section!

File: _summaries/wang2018dynamic_1.md [hunk 4]
```diff
@@ -1,19 +1,95 @@
...
+![Edge Conv](wang2018dynamic_1c.png)
+  - Look at how $$e_{ij}$$ is computed, you take both vertices and feed their features (in this case X,Y,Z coordinates) into linear layers that will be convoled over later
+
+
+// TODO - put the formal definition of EdgeConv here
+
+
+## Okay... but why is it called "Dynamic Graph"
+
+- Once EdgeConv has ran on the point cloud, you now have a bunch of new values for the point cloud that have been mutated.
```
Comment (jaywhang) [inline _summaries/wang2018dynamic_1.md:41]: Maybe I missed something, but I couldn't find any information about how slow this repeated kNN procedure is.  I was worried that, given how DGCNN works best on dense point sets and every layer requires running a kNN, the overall training/inference time would become prohibitively slow.   If you add something about the computational cost of DGCNN, that'd be helpful as additional context.

File: _summaries/wang2018dynamic_1.md [hunk 5]
```diff
@@ -1,19 +1,95 @@
...
+![K Hyperparameter can hurt performance](wang2018dynamic_1g.png)
+
+- The hyperparameter K (how large are the neighbourhoods) is fickle, and can lead to impaired performance if too high
 
 ## TL;DR
-* Three
-* Bullets
-* To highlight the core concepts
+* EdgeConv takes 1 vertex, then compares it through some metric to a bunch of other vertices, which then go through a nonlinear transform then is aggregated with the other neighbouring vertices.
+* Recomputing graph based CNNs can help improve accuracy
```
Comment (jaywhang) [inline _summaries/wang2018dynamic_1.md:94]: Slightly weird wording here -- maybe something like "Recomputing kNN for each layer based on newly computed point vectors can help improve final accuracy"?

File: _summaries/wang2018dynamic_1.md [hunk 6]
```diff
@@ -1,19 +1,95 @@
...
+
+![K Hyperparameter can hurt performance](wang2018dynamic_1g.png)
+
+- The hyperparameter K (how large are the neighbourhoods) is fickle, and can lead to impaired performance if too high
 
 ## TL;DR
-* Three
-* Bullets
-* To highlight the core concepts
+* EdgeConv takes 1 vertex, then compares it through some metric to a bunch of other vertices, which then go through a nonlinear transform then is aggregated with the other neighbouring vertices.
```
Comment (jaywhang) [inline _summaries/wang2018dynamic_1.md:93]: Maybe this point can be a bit more detailed, since other point cloud methods could also compute "some metric to a bunch of other vertices".

File: _summaries/wang2018dynamic_1.md (hunk 1 above)
Comment (Zayne-sprague) [reply _summaries/wang2018dynamic_1.md:20]: Ah right, so, it's windowed self-attention (or at least that's what I thought it sounded like).  Mostly because it is selecting a subset of nodes, then applying edges to every node within that neighbourhood (including itself).  That sounded like Self-Attention (mostly because it was assigning an edge weight to itself as well as everything else in the neighbourhood)

Comment (Zayne-sprague) [reply _summaries/wang2018dynamic_1.md:20]: This was a bit of stretch, if you think it's too loose of a connection I can scrap it.

File: _summaries/wang2018dynamic_1.md (hunk 4 above)
Comment (Zayne-sprague) [reply _summaries/wang2018dynamic_1.md:41]: Sure! I'll try to find something on it, I don't know if it was in the paper either (or I glossed over it)

File: _summaries/wang2018dynamic_1.md (hunk 5 above)
Comment (Zayne-sprague) [reply _summaries/wang2018dynamic_1.md:94]: That sounds much clearer, thanks!

File: _summaries/wang2018dynamic_1.md (hunk 6 above)
Comment (Zayne-sprague) [reply _summaries/wang2018dynamic_1.md:93]: "EdgeConv computes all edges from one vertex to a neighborhood of other vertices, including an edge to the same vertex, these edge values are then aggregated updating the original value of the vertex."

I reworded it.  I am trying to leave out a specific metric because they say you could really use anything.  Dot Product, Euclidean distances, etc. that's how you can end up recreating pointnet/pointnet++ with their implementation.  But it's hard to really sum that up in one nice bullet point.  I think the updated wording gets closer though

File: _summaries/wang2018dynamic_1.md (hunk 3 above)
Comment (Zayne-sprague) [reply _summaries/wang2018dynamic_1.md:50]: Thanks!

File: _summaries/wang2018dynamic_1.md (hunk 4 above)
Comment (Zayne-sprague) [reply _summaries/wang2018dynamic_1.md:41]: Oh I did put a bullet point in the summary about timing

"DGCNN is faster than PointNet++ (slower than PointNet)"

it's in the results section.  I'll add this to the "dynamic graph" section though instead.  (I agree that it fits better here).

File: _summaries/wang2018dynamic_1.md (hunk 2 above)
Comment (Zayne-sprague) [reply _summaries/wang2018dynamic_1.md:33]: Sounds good, I'll add Table 1 here

**Comment by Zayne-sprague** [comment]
> > I need to add the formal definition of EdgeConv in Latex, but I also wonder if I should reduce the size of the images. I took the via screen shots on my monitor because the ones online were too small, but now they are suuuuuper large (but maybe that's a good thing for presenting?)
> 
> You can control the width of an image by manually inserting the image via html:
> 
> ```
> <p align="center">
>   Some alt text
> </p>
> ```

This helped me out a ton! Thank you so much for this :)

**Comment by Zayne-sprague** [comment]
Thank you for the review.  I tried to make sure I addressed every comment and made sure to include your suggestions.

**Review by zhaoyue-zephyrus — APPROVED** [review]
